Document branch-off in greater details (#40)
NixOS/release-wiki: src/Branch-Off.md

ago.

## Jobset creation

The [infrastructure team](https://matrix.to/#/#infra:nixos.org) can create the necessary jobsets
in advance. They will not evaluate because the release branch doesn't exist but it allows the
infrastructure team and the release team to work more asynchronously. Reach out to them 1-2 days
before release to create the necessary Hydra Jobsets. You can link them this section.

- [https://hydra.nixos.org/project/nixos](https://hydra.nixos.org/project/nixos)
    - release-21.05	NixOS 21.05
    - release-21.05-aarch64	NixOS 21.05
    - release-21.05-small	NixOS 21.05
- [https://hydra.nixos.org/project/nixpkgs](https://hydra.nixos.org/project/nixpkgs)
    - nixpkgs-21.05-darwin
    - staging-next-21.05

For example: [21.05 configuration](https://hydra.nixos.org/jobset/nixos/release-21.05#tabs-configuration)

|Field|Value|
|-|-|
|State|Enabled|
|Description|NixOS 21.05 release branch|
|Nix expression|`nixos/release-combined.nix` in input `nixpkgs`|
|Check interval|86400|
|Scheduling shares|500|
|Number of evaluations to keep|1|

Inputs:

|Input name|Type|Values|
|-|-|-|
|`nixpkgs`|Git checkout|`https://github.com/NixOS/nixpkgs.git release-21.05`|

> |`stableBranch`|Boolean|`false`|

Reviewer: When is `stableBranch` appropriate?

Author: It's part of the next chapter. The value will be set to `true` 1-2 days before the actual release because at the time of jobset creation, the release is not official but in beta

# Branch-off

> For these steps "21.05" represents the current release tag and "21.11" represents the next
> (6 months in the future) release tag. "20.11" is the last release which was released 6 months
> ago.

Reviewer: ```suggestion
For these steps "21.05" represents the current release tag and "21.11" represents the next
(6 months in the future) release tag. "20.11" is the last release which was released 6 months
ago.
```

Create these (or similar) variables and use them in all commands below.

```
export VERYOLD=20.11
export OLD=21.05
export NEW=21.11
```

Author: more like `PREVVER`, `CURVER`, `NEXTVER`? Because what you're referring to as `NEW` is in fact not the version that is being released but rather next one.

   ```shell
   git rev-list --count release-21.05
   ```

7. Commit the changes from the previous steps

8. Create the staging branches
   ```shell
   git branch staging-21.05
   git branch staging-next-21.05
   ```

9. Tag the release and push everything

   ```shell
   git tag --annotate --message="Release 21.05-beta" 21.05-beta
   git push upstream release-21.05 21.05-beta staging-21.05 staging-next-21.05
   ```

10. Switch back to the master branch and [increment the `.version`
    file](https://github.com/NixOS/nixpkgs/commit/01268fda85b7eee4e462c873d8654f975067731f#diff-2bc0e46110b507d6d5a344264ef15adaR1)

   ```sh
   echo -n "21.11" > .version
   ```

11. [Update `codeName` in
    `lib/trivial.nix`](https://github.com/NixOS/nixpkgs/commit/01268fda85b7eee4e462c873d8654f975067731f#diff-03f3d41b68f62079c55001f1a1c55c1dR137)
   This will be the name for the next release.

12. [Create a new release notes file for the upcoming release +
    1](https://github.com/NixOS/nixpkgs/commit/01268fda85b7eee4e462c873d8654f975067731f#diff-e7ee5ff686cdcc513ca089d6e5682587R11),
   in our case this is `rl-2111.section.md`. Don't forget to link the new file in the parent markdown file

> 13. Commit the changes

Author: Hmm, doesn't seem like this was done in one commit previously. Will create an example commit in the next days and link it ;)

> 14. Tag master branch so that `git describe` shows the correct metadata.

Author: Huh? Pretty much this:
```
$ git describe master
22.05-pre-45335-g7af10ab6602
```

> 14. Tag master branch so that `git describe` shows the correct metadata.

Reviewer: Then add it as an example!

    ```shell
    git tag --annotate 21.11-pre
    git push upstream master 21.11-pre
    ```

15. Trigger an evaluation of the new jobsets in Hydra

16. Create a channel at `https://nixos.org/channels` by creating a PR to

>     `nixos-org-configurations`, changing [`channels.nix`](https://github.com/NixOS/nixos-org-configurations/blob/master/channels.nix) and notify the infrastructure team

Author: I expanded the text a bit and will add a reference PR once I created the 22.05 one

> 17. Create the `backport staging-21.05` and `backport staging-21.05` [labels](https://github.com/NixOS/nixpkgs/labels) with a description of `Backport PR automatically` and `#0fafaa` as the color

Reviewer: ```suggestion
17. Create the backport [labels](https://github.com/NixOS/nixpkgs/labels) for all new branches:
  -  `backport release-21.05`
  -  `backport staging-21.05` 
  
  Use the description `Backport PR automatically` and  the color value `#0fafaa`.
```

> 18. Inform the [Marketing team](https://matrix.to/#/#marketing:nixos.org) about the upcoming release

Reviewer: They should be aware due to the release calendar. What is the intent here?

Author: I talked to Aaron about this and he told me to better remind them because there is no GO/NO-GO meeting anymore where they would get a reminder that the next release will be out soon.

1. Wait for the staging team to merge the final staging-next iteration

> 1. Fetch and check out the master branch

Reviewer: Waiting for the staging team was and is still marked step 1.

Reviewer: Oh no, you broke the numbering with some weird editor :D

Reviewer: Weird numbering, but looks fine when rendered.

Reviewer 2: It's likely that even further in future there will be changes in this list, and it would be quite a bother to keep the numbers consistent in the source.

    - Commit the changelog after running `md-to-db.sh`

> 1. Update the [periodic-merge workflow](https://github.com/NixOS/nixpkgs/blob/master/.github/workflows/periodic-merge-24h.yml) to include the new branch and commit + push master

Reviewer: ```suggestion
1. Update the [periodic-merge workflow](https://github.com/NixOS/nixpkgs/blob/master/.github/workflows/periodic-merge-24h.yml) to include the new branch, then commit & push to master
```
